pkg/common/utils/utils_test.go

```diff
@@ -736,6 +739,443 @@ func TestNeedRecovery(t *testing.T) {
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pod-01",
+				UID:  "UID-01",
+				Annotations: map[string]string{
+					constants.AnnotationPlaceholderFlag: "true",
+				},
+			},
+		}, true},
+		{"Setting by label", &v1.Pod{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Pod",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pod-01",
+				UID:  "UID-01",
+				Labels: map[string]string{
+					constants.LabelPlaceholderFlag: "true",
+				},
+			},
+		}, true},
+		{"Setting both annotation and label, annotation has higher priority", &v1.Pod{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Pod",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pod-01",
+				UID:  "UID-01",
+				Annotations: map[string]string{
+					constants.AnnotationPlaceholderFlag: "true",
+					constants.LabelPlaceholderFlag:      "false",
+				},
+			},
+		}, true},
+		{"No setting both annotation and label", &v1.Pod{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Pod",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pod-01",
+				UID:  "UID-01",
+			},
+		}, false},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			placeHolderFlag := GetPlaceholderFlagFromPodSpec(tc.pod)
+			assert.Equal(t, placeHolderFlag, tc.expectedPlaceholderFlag)
+		})
+	}
+}
+
+// nolint: funlen
+func TestGetTaskGroupFromAnnotation(t *testing.T) {
+	// correct json
+	testGroup := `
+	[
+		{
+			"name": "test-group-1",
+			"minMember": 10,
+			"minResource": {
+				"cpu": 1,
+				"memory": "2Gi"
+			},
+			"nodeSelector": {
+				"test": "testnode",
+				"locate": "west"
+			},
+			"tolerations": [
+				{
+					"key": "key",
+					"operator": "Equal",
+					"value": "value",
+					"effect": "NoSchedule"
+				}
+			]
+		},
+		{
+			"name": "test-group-2",
+			"minMember": 5,
+			"minResource": {
+				"cpu": 2,
+				"memory": "4Gi"
+			}
+		}
+	]`
+	testGroup2 := `
+	[
+		{
+			"name": "test-group-3",
+			"minMember": 3,
+			"minResource": {
+				"cpu": 2,
+				"memory": "1Gi"
+			}
+		}
+	]`
+	// Error json
+	testGroupErr := `
+	[
+		{
+			"name": "test-group-err-1",
+			"minMember": "ERR",
+			"minResource": {
+				"cpu": "ERR",
+				"memory": "ERR"
+			},
+		}
+	]`
+	// without name
+	testGroupErr2 := `
+	[
+		{
+			"minMember": 3,
+			"minResource": {
+				"cpu": 2,
+				"memory": "1Gi"
+			}
+		}
+	]`
+	// without minMember
+	testGroupErr3 := `
+	[
+		{
+			"name": "test-group-err-2",
+			"minResource": {
+				"cpu": 2,
+				"memory": "1Gi"
+			}
+		}
+	]`
+	// without minResource
+	testGroupErr4 := `
+	[
+		{
+			"name": "test-group-err-4",
+			"minMember": 3
+		}
+	]`
+	// negative minMember without minResource
+	testGroupErr5 := `
+	[
+		{
+			"name": "test-group-err-5",
+			"minMember": -100
+		}
+	]`
+	// negative minMember with minResource
+	testGroupErr6 := `
+	[
+		{
+			"name": "test-group-err-5",
+			"minMember": -100
```

Comment:
Miss comma.
After adding comma, this case would classified to the minMembercase for improving coverage.
```
"minMember": -100,
```

Reply:
You can view the coverage of codes in web.
Run the "go tool cover -html coverage.txt" after running the "make test"

Reply:
Good catch! Thank you, @0yukali0, for patient review, addressed in latest PR.